rediscluster/pipeline.py

```diff
@@ -142,144 +146,70 @@ def send_cluster_commands(self, stack, raise_on_error=True, allow_redirections=T
+
+        # as we move through each command that still needs to be processed,
+        # we figure out the slot number that command maps to, then from the slot determine the node.
+        for c in attempt:
+            # refer to our internal node -> slot table that tells us where a given
+            # command should route to.
+            slot = self._determine_slot(*c.args)
+            node = self.connection_pool.get_node_by_slot(slot)
+
+            # little hack to make sure the node name is populated. probably could clean this up.
+            self.connection_pool.nodes.set_node_name(node)
+
+            # now that we know the name of the node ( it's just a string in the form of host:port )
+            # we can build a list of commands for each node.
+            node_name = node['name']
+            if node_name not in nodes:
+                nodes[node_name] = NodeCommands(self.parse_response, self.connection_pool.get_connection_by_node(node))
+
+            nodes[node_name].append(c)
+
+        # send the commands in sequence.
+        # we  write to all the open sockets for each node first, before reading anything
+        # this allows us to flush all the requests out across the network essentially in parallel
+        # so that we can read them all in parallel as they come back.
+        # we dont' multiplex on the sockets as they come available, but that shouldn't make too much difference.
+        node_commands = nodes.values()
+        for n in node_commands:
+            n.write()
+
+        for n in node_commands:
+            n.read()
+
+        # release all of the redis connections we allocated earlier back into the connection pool.
+        # we used to do this step as part of a try/finally block, but it is really dangerous to
+        # release connections back into the pool if for some reason the socket has data still left in it
+        # from a previous operation. The write and read operations already have try/catch around them for
+        # all known types of errors including connection and socket level errors.
+        # So if we hit an exception, something really bad happened and putting any of
+        # these connections back into the pool is a very bad idea.
+        # the socket might have unread buffer still sitting in it, and then the
+        # next time we read from it we pass the buffered result back from a previous
+        # command and every single request after to that connection will always get
+        # a mismatched result. (not just theoretical, I saw this happen on production x.x).
+        for n in nodes.values():
+            self.connection_pool.release(n.connection)
+
+        # if the response isn't an exception it is a valid response from the node
+        # we're all done with that command, YAY!
+        # if we have more commands to attempt, we've run into problems.
+        attempt = sorted([c for c in attempt if isinstance(c.result, ERRORS_ALLOW_RETRY)], key=lambda x: x.position)
```

Comment:
Can you think of a better name for this variable `attempt` because it do not feel like an attempt any longer as the logic for the attempts was moved to other places.